File: src/lib/tariff.ts
/**
 * Единый источник истины для тарификации грузов CargoTrans.
 * Используется во всех компонентах: CargoDetails, Payment, NewShipment.
 * Также синхронизирован с backend/internal/service/tariff.go → calculateCostByTariff()
 */

/** Компоненты тарифа для маршрута (за 10 кг) */
interface RouteTariff {
  transportRate: number;       // Тариф за перевозку (за 10 кг)
  declaredValueFee: number;    // Объявленная ценность (за 10 кг)
}

/** Тарифы по маршрутам. Ключ — "откуда-куда" в нижнем регистре. */
export const ROUTE_TARIFFS: Record<string, RouteTariff> = {
  'алматы-2-астана нұрлы жол': { transportRate: 971, declaredValueFee: 84 },
  'астана нұрлы жол-алматы-2': { transportRate: 971, declaredValueFee: 84 },
  'алматы-2-қарағанды':       { transportRate: 819, declaredValueFee: 68 },
  'қарағанды-алматы-2':       { transportRate: 819, declaredValueFee: 68 },
};

/** Fallback тариф для нераспознанных маршрутов (9769 за 100кг = 976.9 за 10кг) */
export const FALLBACK_TARIFF: RouteTariff = { transportRate: 976.9, declaredValueFee: 0 };

/** Накладная — фиксированный сбор за распечатывание (₸) */
export const WAYBILL_FEE = 107;

/** Скидка при наличии ж/д билета */
export const TICKET_DISCOUNT = 0.5;

export interface TariffParams {
  fromStation: string;
  toStation: string;
  weight: string | number;
  hasTicket?: boolean;
  isDoorToDoor?: boolean;
  clientType?: string;
}

/**
 * Возвращает тариф для маршрута.
 */
export function getRouteTariff(from: string, to: string): RouteTariff {
  const key = `${from.trim().toLowerCase()}-${to.trim().toLowerCase()}`;
  return ROUTE_TARIFFS[key] || FALLBACK_TARIFF;
}

/**
 * Округляет вес вверх до ближайших 10 кг (минимум 10 кг).
 */
export function roundWeight(weight: number): number {
  const rounded = Math.ceil(weight / 10) * 10;
  return Math.max(rounded, 10);
}

/**
 * Возвращает детализацию расчёта стоимости.
 */
export function getCostBreakdown(params: TariffParams) {
  const { fromStation, toStation, weight, hasTicket, isDoorToDoor, clientType } = params;

  if (!fromStation || !toStation || !weight) return null;

  const weightNum = typeof weight === 'string' ? parseFloat(weight) : weight;
  if (isNaN(weightNum) || weightNum <= 0) return null;

  const roundedWeight = roundWeight(weightNum);
  const blocks = roundedWeight / 10;
  const tariff = getRouteTariff(fromStation, toStation);

  const transportCost = blocks * tariff.transportRate;
  const declaredValueCost = tariff.declaredValueFee;
  let subtotal = transportCost + declaredValueCost;

  if (hasTicket) subtotal *= TICKET_DISCOUNT;

  let doorToDoorCost = 0;
  if (isDoorToDoor && clientType !== 'legal') {
    doorToDoorCost = 10000;
  }

  const total = Math.round(subtotal + doorToDoorCost + WAYBILL_FEE);

  return {
    weightNum,
    roundedWeight,
    blocks,
    tariff,
    transportCost: Math.round(transportCost),
    declaredValueCost: Math.round(declaredValueCost),
    doorToDoorCost,
    waybillFee: WAYBILL_FEE,
    hasTicketDiscount: !!hasTicket,
    total,
  };
}

/**
 * Рассчитывает итоговую стоимость перевозки.
 * Формула: ceil(вес/10) × (перевозка + обьявл.ценность) + накладная + надбавки
 * Возвращает null если данных недостаточно.
 */
export function calculateShipmentCost(params: TariffParams): number | null {
  const breakdown = getCostBreakdown(params);
  return breakdown ? breakdown.total : null;
}

File: backend/internal/service/tariff.go

package service

import (
	"fmt"
	"math"
	"strings"
)

// routeTariff содержит компоненты тарифа для маршрута (за 10 кг)
type routeTariff struct {
	TransportRate    float64 // Тариф за перевозку (за 10 кг)
	DeclaredValueFee float64 // Объявленная ценность (за 10 кг)
}

// Тарифы по маршрутам. Ключ — "откуда-куда" в нижнем регистре.
var routeTariffs = map[string]routeTariff{
	"алматы-2-астана нұрлы жол": {TransportRate: 971.0625, DeclaredValueFee: 84},
	"астана нұрлы жол-алматы-2": {TransportRate: 971.0625, DeclaredValueFee: 84},
	"алматы-2-қарағанды":       {TransportRate: 819, DeclaredValueFee: 68},
	"қарағанды-алматы-2":       {TransportRate: 819, DeclaredValueFee: 68},
	"астана нұрлы жол-қарағанды": {TransportRate: 287, DeclaredValueFee: 16},
	"қарағанды-астана нұрлы жол": {TransportRate: 287, DeclaredValueFee: 16},
}

// Накладная — фиксированный сбор за распечатывание (₸)
const waybillFee = 107

func calculateCostByTariff(fromStation, toStation string, weightStr string, description string, isDoorToDoor bool, isIndividual bool) float64 {
	if fromStation == "" || toStation == "" || weightStr == "" {
		return 0
	}

	var weight float64
	_, err := fmt.Sscanf(weightStr, "%f", &weight)
	if err != nil || weight <= 0 {
		return 0
	}

	// Округляем вес вверх до ближайших 10 кг (минимум 10 кг)
	roundedWeight := math.Ceil(weight/10.0) * 10
	if roundedWeight < 10 {
		roundedWeight = 10
	}
	blocks := roundedWeight / 10

	// Ищем тариф по маршруту
	routeKey := strings.TrimSpace(strings.ToLower(fromStation)) + "-" + strings.TrimSpace(strings.ToLower(toStation))
	tariff, found := routeTariffs[routeKey]
	if !found {
		// Если маршрут не найден — fallback на старую логику (9769 за 100кг = 976.9 за 10кг)
		tariff = routeTariff{TransportRate: 976.9, DeclaredValueFee: 0}
	}

	cost := (blocks * tariff.TransportRate) + tariff.DeclaredValueFee

	// +10 000 тг для door-to-door от физлица (Фаза 5, status_logic.md)
	if isDoorToDoor && isIndividual {
		cost += 10000
	}

	// +107 тг за распечатывание накладной
	cost += float64(waybillFee)

	return math.Round(cost)
}
